src/uu/echo/src/echo.rs

```diff
@@ -23,63 +23,104 @@ mod options {
-                    result.push(OsString::from("--"));
-                }
+/// Checks if an argument is a valid echo flag
+/// Returns true if valid echo flag found
+fn is_echo_flag(arg: &OsString, echo_options: &mut EchoOptions) -> bool {
+    let bytes = arg.as_encoded_bytes();
+    if bytes.first() == Some(&b'-') && arg != "-" {
+        // we initialize our local variables to the "current" options so we don't override
+        // previous found flags
+        let mut escape = echo_options.escape;
+        let mut trailing_newline = echo_options.trailing_newline;
+
+        // Process characters after the '-'
+        for c in &bytes[1..] {
+            match c {
+                b'e' => escape = true,
+                b'E' => escape = false,
+                b'n' => trailing_newline = false,
+                // if there is any char in an argument starting with '-' that doesn't match e/E/n
+                // present means that this argument is not a flag
+                _ => return false,
             }
-            result.push(arg);
         }
+
+        // we only override the options with flags being found once we parsed the whole argument
+        echo_options.escape = escape;
+        echo_options.trailing_newline = trailing_newline;
+        return true;
     }
 
-    result.into_iter()
+    // argument doesn't start with '-' or is "-" => no flag
+    false
 }
 
-fn collect_args(matches: &ArgMatches) -> Vec<OsString> {
-    matches
-        .get_many::<OsString>(options::STRING)
-        .map_or_else(Vec::new, |values| values.cloned().collect())
+/// Processes command line arguments, separating flags from normal arguments
+/// Returns:
+/// - Vector of non-flag arguments
+/// - trailing_newline: whether to print a trailing newline
+/// - escape: whether to process escape sequences
+fn filter_echo_flags(args: impl uucore::Args) -> (Vec<OsString>, bool, bool) {
+    let mut result = Vec::new();
+    let mut echo_options = EchoOptions {
+        trailing_newline: true,
+        escape: false,
+    };
+    let mut args_iter = args.into_iter();
+
+    // Process arguments until first non-flag is found
+    for arg in &mut args_iter {
+        // we parse flags and store options found in "echo_option". First is_echo_flag
+        // call to return false will break the loop and we will collect the remaining arguments
+        if !is_echo_flag(&arg, &mut echo_options) {
+            // First non-flag argument stops flag processing
+            result.push(arg);
+            break;
+        }
+    }
+    // Collect remaining arguments
+    for arg in args_iter {
+        result.push(arg);
+    }
+    (result, echo_options.trailing_newline, echo_options.escape)
 }
 
 #[uucore::main]
 pub fn uumain(args: impl uucore::Args) -> UResult<()> {
-    let is_posixly_correct = env::var("POSIXLY_CORRECT").is_ok();
+    // Check POSIX compatibility mode
+    let is_posixly_correct = env::var_os("POSIXLY_CORRECT").is_some();
```

Comment:
There is a subtle change in behavior here: `POSIXLY_CORRECT=<non-utf8-sequence>` is now interpreted as `true`, whereas before it was interpreted as `false`. I assume this is a correct change. (I can't test GNU easily right now.)

```console
$ POSIXLY_CORRECT=$'\377asd' cargo run echo 'asd\nqwe' # current main
asd\nqwe
$ POSIXLY_CORRECT=$'\377asd' cargo run echo 'asd\nqwe' # your PR
asd
qwe
```

Nitpick: If you fix subtleties like this, it would be extra awesome if you also add a test, so that we don't lose it later :)

Reply:
Awesome! But I would argue that if you change the behavior, then an old test cannot possibly correctly test for the new behavior :D

In particular, this test only tests whether `-e` is interpreted or considered as part of the output. It does not test whether escape sequences are interpreted or taken literally, i.e. whether "escape mode" is initially set or not.

Like I said, it's not a major thing, and doesn't block this PR.

Reply:
It's not an old test, I meant that was where I'd put it. It fails if `env::var(...).is_ok()` is used.